File: 1143-longest-common-subsequence/1143-longest-common-subsequence.cpp
class Solution {
public:
int helper(int idx1,int idx2,string str1,string str2, vector<vector<int>>&dp){

    if(idx1<0 || idx2<0) return 0;
    if(dp[idx1][idx2] != -1) return dp[idx1][idx2];
    int match =0;
    if(str1[idx1] == str2[idx2]){
        match = 1 + helper(idx1-1,idx2-1,str1,str2,dp);
    }

    int notmatch = max(helper(idx1-1,idx2,str1,str2,dp), helper(idx1,idx2-1,str1,str2,dp));

    return dp[idx1][idx2] = max(match , notmatch);
}
    int longestCommonSubsequence(string text1, string text2) {
        
        vector<vector<int>>dp(text1.size()+1,vector<int>(text2.size()+1,0));
        //return helper(text1.size()-1,text2.size()-1,text1,text2,dp);

        // for(int i=0;i<text1.size();i++){
        //     dp[0][i] = 0;
        // }
        // for(int i=0;i<text2.size();i++){
        //     dp[i][0] = 0;
        // }

        for(int i=1;i<=text1.size();i++){
            for(int j=1;j<=text2.size();j++){

                int match = 0;
                if(text1[i-1] == text2[j-1]){
                    match = 1 + dp[i-1][j-1];
                }
                int notmatch = max(dp[i-1][j],dp[i][j-1]);

                dp[i][j] = max(match, notmatch);
            }
        }

        return dp[text1.size()][text2.size()];
    }
};

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {

    int help(int idx1, int idx2,String s1, String s2, int[][] dp){
       if(idx1 < 0 || idx2 < 0) return 0;


       if(dp[idx1][idx2] != -1) return dp[idx1][idx2];

       if(s1.charAt(idx1) == s2.charAt(idx2)){
        return dp[idx1][idx2] = 1 + help(idx1-1,idx2-1,s1,s2,dp);
       }
       return dp[idx1][idx2] = Math.max(help(idx1-1,idx2,s1,s2,dp),help(idx1,idx2-1,s1,s2,dp));

    }
    public int longestCommonSubsequence(String text1, String text2) {
        int[][] dp = new int[text1.length()+1][text2.length()+1];

        for(int[] a: dp){
         Arrays.fill(a,-1);
        }
       return help(text1.length()-1,text2.length()-1,text1,text2,dp);
      
    }
}